executor/index_merge_reader.go

```diff
@@ -775,7 +789,157 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
 	}
 }
 
-func (w *indexMergeProcessWorker) handleLoopFetcherPanic(ctx context.Context, resultCh chan<- *lookupTableTask) func(r interface{}) {
+type intersectionProcessWorker struct {
+	workerID int
+	// key: parTblIdx, val: HandleMap
+	handleMapsPerWorker map[int]*kv.MemAwareHandleMap[*int]
+	workerCh            chan *indexMergeTableTask
+	indexMerge          *IndexMergeReaderExecutor
+	wg                  *sync.WaitGroup
+	memTracker          *memory.Tracker
+	memUsage            int64
+}
+
+func (w *intersectionProcessWorker) doIntersectionPerPartition() {
+	for task := range w.workerCh {
+		var ok bool
+		var hMap *kv.MemAwareHandleMap[*int]
+		if hMap, ok = w.handleMapsPerWorker[task.parTblIdx]; !ok {
+			panic(fmt.Sprintf("cannot find parTblIdx(%d) for worker(id: %d)", task.parTblIdx, w.workerID))
+		}
+		var mapDelta int64
+		var rowDelta int64
+		for _, h := range task.handles {
+			if cntPtr, ok := hMap.Get(h); ok {
+				(*cntPtr)++
+			} else {
+				cnt := 1
+				mapDelta += hMap.Set(h, &cnt) + int64(h.ExtraMemSize())
+				rowDelta += 1
+			}
+		}
+		if rowDelta > 0 {
+			memDelta := mapDelta + (rowDelta * int64(unsafe.Sizeof(reflect.Pointer)))
+			w.memUsage += memDelta
+			w.memTracker.Consume(mapDelta)
+		}
+	}
+	w.wg.Done()
+}
+
+func (w *indexMergeProcessWorker) fetchLoopIntersection(ctx context.Context, fetchCh <-chan *indexMergeTableTask,
+	workCh chan<- *indexMergeTableTask, resultCh chan<- *indexMergeTableTask, finished <-chan struct{}) {
+	defer func() {
+		close(workCh)
+		close(resultCh)
+	}()
+
+	if w.stats != nil {
+		start := time.Now()
+		defer func() {
+			w.stats.IndexMergeProcess += time.Since(start)
+		}()
+	}
+
+	partCnt := 1
+	workerCnt := 1
+	partCntPerWorker := 1
+	// One goroutine may handle one or multiple partitions.
+	// Max number of partition number is 8192, we use ExecutorConcurrency to avoid too many goroutines.
+	maxWorkerCnt := w.indexMerge.ctx.GetSessionVars().IndexMergeIntersectionConcurrency()
+	maxChannelSize := 1024
+	if w.indexMerge.partitionTableMode {
+		partCnt = len(w.indexMerge.prunedPartitions)
+		workerCnt = partCnt
+		if workerCnt > maxWorkerCnt {
+			workerCnt = maxWorkerCnt
+			partCntPerWorker = partCnt / workerCnt
+			if partCnt%workerCnt != 0 {
+				partCntPerWorker++
+			}
+		}
+	}
+	failpoint.Inject("testIndexMergeIntersectionConcurrency", func(val failpoint.Value) {
+		con := val.(int)
+		if con != workerCnt {
+			panic(fmt.Sprintf("unexpected workerCnt, expect %d, got %d", con, workerCnt))
+		}
+	})
+	handleMaps := make([]*kv.MemAwareHandleMap[*int], 0, partCnt)
+	workers := make([]*intersectionProcessWorker, 0, workerCnt)
+	wg := sync.WaitGroup{}
```

Comment:
We can use `util.WaitGroupWrapper`

Reply:
And we need a recover function for the goroutine

Reply:
Done

Reply:
We can use 
```
wg := util.WaitGroupWrapper{}
wg.RunWithRecover()
wg.Wait()
```

Reply:
done